src/main/java/com/artipie/jfr/JfrStorage.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * The MIT License (MIT) Copyright (c) 2020-2021 artipie.com
+ * https://github.com/artipie/artipie/LICENSE.txt
+ */
+package com.artipie.jfr;
+
+import com.artipie.asto.Content;
+import com.artipie.asto.Key;
+import com.artipie.asto.Meta;
+import com.artipie.asto.Storage;
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+
+/**
+ * Wrapper for a Storage that generates JFR events for operations.
+ *
+ * @since 0.28.0
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
+ */
+public final class JfrStorage implements Storage {
+
+    /**
+     * Original storage.
+     */
+    private final Storage original;
+
+    /**
+     * Ctor.
+     *
+     * @param original Original storage.
+     */
+    public JfrStorage(final Storage original) {
+        this.original = original;
+    }
+
+    @Override
+    public CompletableFuture<Boolean> exists(final Key key) {
+        final StorageExistsEvent event = new StorageExistsEvent();
+        event.storage = this.identifier();
+        event.key = key.string();
+        return this.original.exists(key)
+            .thenApply(
+                res -> {
+                    event.commit();
+                    return res;
+                }
+            );
+    }
+
+    @Override
+    public CompletableFuture<Void> save(final Key key, final Content content) {
+        final StorageSaveEvent event = new StorageSaveEvent();
+        event.storage = this.identifier();
+        event.key = key.string();
+        event.begin();
+        try {
+            return this.original.save(
+                key,
+                new ChunksAndSizeMetricsContent(
+                    content,
+                    (chunks, size) -> {
+                        event.chunks = chunks;
+                        event.size = size;
+                        event.commit();
+                    }
+                )
+            );
+        } finally {
+            event.commit();
+        }
```

Comment:
is it correct to call `commit()` twice? it will first be called in finally-block and when publisher is fully consumed...

Reply:
if we want to commit the event in the case of error it will be better to use `handle()` method

Reply:
I forget to delete this after debug ((